pkg/odo/cli/component/create.go

```diff
@@ -121,324 +113,118 @@ func NewCreateOptions() *CreateOptions {
-	// we only get absolute path for local source type
-	if sourcePath == "" {
-		wd, err := os.Getwd()
-		if err != nil {
-			return "", err
-		}
-		return wd, nil
-	}
-	finalSourcePath, err := filepath.Abs(sourcePath)
-	if err != nil {
-		return "", err
-	}
-	return finalSourcePath, nil
-
-}
-
-// Complete completes create args
 func (co *CreateOptions) Complete(name string, cmd *cobra.Command, args []string) (err error) {
-
-	if co.now {
-		// this populates the EnvInfo as well
-		co.Context, err = genericclioptions.NewContextCreatingAppIfNeeded(cmd)
-		if err != nil {
-			return err
-		}
-	} else {
-		co.Context, err = genericclioptions.NewOfflineContext(cmd)
-		if err != nil {
-			return err
-		}
+	// GETTERS
+	// Get context
+	co.Context, err = getContext(co.now, cmd)
+	if err != nil {
+		return err
 	}
-
-	DevfilePath := location.DevfileLocation("")
-	// Configure the context
-	if co.componentContext != "" {
-		DevfilePath = location.DevfileLocation(co.componentContext)
-		EnvFilePath = filepath.Join(co.componentContext, envFile)
-		co.PushOptions.componentContext = co.componentContext
+	// Get the app name
+	co.appName = genericclioptions.ResolveAppFlag(cmd)
+	// Get the project name
+	co.devfileMetadata.componentNamespace = co.Context.Project
+	// Get DevfilePath
+	co.DevfilePath = location.DevfileLocation(co.componentContext)
+	//Check whether the directory already contains a devfile, this check should happen early
+	co.devfileMetadata.userCreatedDevfile = util.CheckPathExists(co.DevfilePath)
+	// EnvFilePath is the path of env file for devfile component
+	envFilePath := getEnvFilePath(co.componentContext)
+	// This is required so that .odo is created in the correct context
+	co.PushOptions.componentContext = co.componentContext
+	// Use Interactive mode if: 1) no args are passed || 2) the devfile exists || 3) --devfile is used
+	if len(args) == 0 && !util.CheckPathExists(co.DevfilePath) && co.devfileMetadata.devfilePath.value == "" {
+		co.interactive = true
 	}
-	co.DevfilePath = DevfilePath
 
-	if util.CheckPathExists(EnvFilePath) && util.CheckPathExists(co.DevfilePath) {
+	// CONFLICT CHECK
+	// Check if a component exists
+	if util.CheckPathExists(envFilePath) && util.CheckPathExists(co.DevfilePath) {
 		return errors.New("this directory already contains a component")
 	}
-
-	if util.CheckPathExists(EnvFilePath) && !util.CheckPathExists(co.DevfilePath) {
+	// Check if there is a dangling env file; delete the env file if found
+	if util.CheckPathExists(envFilePath) && !util.CheckPathExists(co.DevfilePath) {
 		log.Warningf("Found a dangling env file without a devfile, overwriting it")
-		if err = util.DeletePath(EnvFilePath); err != nil {
+		// Note: if the IF condition seems to have a side-effect, it is better to do the condition check separately, like below
+		err = util.DeletePath(envFilePath)
+		if err != nil {
 			return err
 		}
 	}
-
+	//Check if the directory already contains a devfile when --devfile flag is passed
```

Comment:
could we move these checks to the implementation level? (in `FetchDevfileAndCreateComponent` or in some `Check` method?)